Customizable priorities of address names sources

tomvodi opened the pull request:
Closes #4487 

## Checklist

- [x] The PR modified the frontend, and updated the [user guide](https://github.com/rotki/rotki/blob/develop/docs/usage_guide.rst) to reflect the changes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/rotki/rotki?pullRequest=5074) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/rotki/rotki?pullRequest=5074) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/rotki/rotki?pullRequest=5074) it.</sub>

LefterisJP reviewed (commented):
Hey @tomvodi thanks for the PR! This needs another rebase on develop

kelsos commented on frontend/app/src/components/settings/frontend/AddressNamePrioritySetting.vue:
```diff
@@ -0,0 +1,73 @@
+        :disable-add="true"
+        :disable-delete="true"
+        :status="{ error, success }"
+        @input="update"
```
A thing that would probably need to be done after changing the priority, is to also re-fetch the names, because, the frontend caches the names and the priority change won't be instantly reflected otherwise.

https://github.com/rotki/rotki/blob/9418c8059c1a162b0b5cd39314e6e6e1e08504d1/frontend/app/src/store/balances/ethereum-names.ts#L82

kelsos reviewed (commented):
Thank you for the PR,

I tested the new component and it works like a charm.

I left some comments about some changes I would like (most of the are styling related), but overall it looks good. Thanks for the nice component test too.

There also seem to be some linter errors.
You should be able auto-fix some of them by running `npm run lint:fix` but it is a bit slow.

I did not see any new warnings, only the ones we have for the number of imports, so on that front things should be fine.

tomvodi commented on frontend/app/src/types/prioritized-list-data.ts:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+export class PrioritizedListItemData {
```
Yes, I had defined it as an interface in the first place, but got an error in `PrioritizedListEntry` where I define the data property as type `PrioritizedListItemData`. I now stumbled upon this `PropType` thingy, so I'll use this :)

tomvodi commented on frontend/app/src/components/helper/PrioritizedList.vue:
```diff
@@ -133,13 +152,21 @@ const props = defineProps({
 const input = (items: string[]) => emit('input', items);
 
+const itemNameTr = computed(() => {
+  let name = get(itemDataName);
```
would be great if the linter catches this

tomvodi commented on frontend/app/src/components/helper/PrioritizedListEntry.vue:
```diff
@@ -0,0 +1,93 @@
+        />
+      </adaptive-wrapper>
+    </v-col>
+    <v-col v-if="data.identifier === PriceOracle.UNISWAP2" cols="auto">
+      {{ t('oracles.uniswap_v2') }}
```
That would look much nicer. The problem I get here is that `data.identifier` has type string and the keys of the `labels` object have the values of the **two** enum types and I get the following error:

`Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ uniswapv2: TranslateResult; uniswapv3: TranslateResult; manualcurrent: TranslateResult; blockchain_account: TranslateResult; ... 4 more ...; private_addressbook: TranslateResult; }'.
  No index signature with a parameter of type 'string' was found on type '{ uniswapv2: TranslateResult; uniswapv3: TranslateResult; manualcurrent: TranslateResult; blockchain_account: TranslateResult; ... 4 more ...; private_addressbook: TranslateResult; }'.`

My preferred solution was to put these translated strings into the `PrioritizedListItemData` class/interface as a property `displayName` for example. This property would be initialized along with the identifier and other stuff like the icons. In the  `PrioritizedListEntry` there is only a `tr(data.displayName)`. This would have the benefit, that this visual item data is defined in one place and the generic `PrioritizedListEntry` component is not aware of the implementations and their translation strings. 
Unfortunately this this result in linter errors complaining about missing translations. I guess the linter is only looking in Vue components and not in .ts scripts. I haven't found a good solution for this after some investigation.

Do you have any solution for this? I must admit, that I'm not that familiar with typescript. I'm more a backend dev (go, C++, a bit of python ;) ) and working with frontend (Angular) for some months now.

kelsos replied:
Let me think a bit about it and look at the code again.

Does `data.identifier` need to be a string? It can only take values from the two enums right?

kelsos replied:
> Let me think a bit about it and look at the code again.
> 
> Does `data.identifier` need to be a string? It can only take values from the two enums right?

Ok I see that it can be more, so what I would do in this case I would go the generics way

```typescript
export class PrioritizedListData<T = string> {
  constructor(private itemData: Array<PrioritizedListItemData<T>>) {}

  itemIdsNotIn(itemIds: T[]): T[] {
    return this.itemData
      .filter(item => !itemIds.includes(item.identifier))
      .map(itemData => itemData.identifier);
  }

  itemDataForId(id: T): PrioritizedListItemData<T> | undefined {
    return this.itemData.find(item => item.identifier === id);
  }
}

export interface PrioritizedListItemData<T = string> {
  identifier: T;
  icon?: string;
  extraDisplaySize?: string; // may be a string specifying pixels e.g. 12px
}
```

and then at the data initialisation I would pass the enum types so that identifier is correctly typed.

```typescript
const props = defineProps({
  data: { required: true, type: Object as PropType<PrioritizedListItemData<AddressNamePriority>> }
});
```

Then use a mapped type for the labels:

```typescript
const labels: { [keys in AddressNamePriority]: string } = {
  [AddressNamePriority.BLOCKCHAIN_ACCOUNT]: t(
    'eth_address_book.hint.priority.list.private_address_book'
  ).toString()
  ...
};
```

This way `{{ labels[data.identifier] }}` should work without any issues.

tomvodi replied:
Maybe I get something wrong here, but in this case I need to specify the `PrioritizedList.vue` and the `PrioritizedListEntry.vue` twice, once with `AddressNamePriority` type and once with `PriceOracle` type? 
The labels object won't take that `PriceOracle` enum anymore :(

kelsos replied:
You no need you can a custom type that is `AddressNamePriority | PriceOracle`

And if you only need specific types from `PriceOracle` you can use [`Pick`](https://www.typescriptlang.org/docs/handbook/utility-types.html#picktype-keys) to narrow down the type. This should include all the properties you want.

kelsos replied:
If it is overly complicated let me know when you are don't with other review comments and I can take it over and push on top of your PR.

tomvodi replied:
I'll try it by myself and will come back to you if I'm stuck. Thanks for your help!

LefterisJP reviewed (commented):
Hey @tomvodi thanks a lot for the PR and for tackling a non-trivial task for your first rotki PR. Something that not only touches frontend but backend!

Left quite a bit of comments so ask any questions either here or in discord if anything is not clear.

Also please add a changelog entry for this since it's a user facing change.

Something like: `Users can now customize the order of resolution with which addresses are resolved to human readable name`

tomvodi commented:
Hi @LefterisJP and @kelsos. Thanks for the feedback!
One question came up as I was fixing the issue @kelsos  mentioned about re-fetching the names after changing the priority: https://github.com/rotki/rotki/pull/5074#discussion_r1012129302.
If I changed the order of name resolution, the Accounts & Balances shows me my addresses the right way. Now, when I restart rotki, the Accounts & Balances show me initially the blockchain account name and not what I set as a higher priority.
I investigated this and I found out that initially the data is fetched from the `get_blockchain_account_data` from the rotkehlchen backend and uses the blockchain label, if set.
This should be the case also for the current version, where the hardcoded priority ist 'private addressbook' -> 'blockchain account label'. 
So my question is, if this is the intended behavior or if the initial loaded data should also show the user-set priority?

kelsos commented:
The naming mechanism was something that was added later, and I think the component reacts to the data, but the application doesn't fetch the data from the get go, so it defaults to the original way which is showing what data the frontend has.

@LefterisJP We need to fix this, and there are two possible approaches.
1. Ignore the names endpoint data and show label/address as before
2. Fetch names on login, and let it show what the names endpoint returns

LefterisJP commented:
> So my question is, if this is the intended behavior or if the initial loaded data should also show the user-set priority?

As @kelsos said the labels were made long time before the whole names concept even existed.

I think this should be changed and the user chosen priority should always take precedence.

From the approaches kelsos mentioned only (2) makes sense

tomvodi commented on rotkehlchen/tests/unit/test_ethereum_names.py:
```diff
@@ -0,0 +1,72 @@
+    return DEFAULT_ADDRESS_NAME_PRIORITY
+
+
+@pytest.fixture(name='fetchers')
```
The linter gave me errors about `Redefining name ... from outer scope (redefined-outer-name)`. This happens because of the fixtures defined in the same module: https://stackoverflow.com/questions/46089480/pytest-fixtures-redefining-name-from-outer-scope-pylint
I've found the mentioned fix in other places in the code, so I decided to do it that way.

tomvodi commented on rotkehlchen/tests/unit/test_ethereum_names.py:
```diff
@@ -0,0 +1,72 @@
+        return self.address_names
+
+
+@pytest.fixture(name='evm_address')
```
I'll use this function, but I still need that fixture, as the function returns a new address for every call

LefterisJP replied:
you don't need to make a fixture for everything. Just save the address in a variable

tomvodi commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +70,171 @@ def search_for_addresses_names(
+        pass
+
+
+class NamePrioritizer:
```
1. Style: For the style issue, is there a definition on when to split imports on multiple lines? I now run the `isort` with the parameters `--wrap-length=99 --line-length=99 --multi-line=3` to have it the way it is styled in other places but it won't catch the 99 columns long import in the `test_ethereum_names.py`.
2. You're right. The ABC is a bit over engineered in this place. I thought the global/private addressbook is also used for BTC and other chains but foud out today that its only used for ETH. My primary intention was to separate the name finding logic from the database access stuff for better testability/SRP.
3. Iterating over the adresses and query every source for that specific address was indeed my first approach. But I took a look at the previous implementation where each data source is queried with all addresses and I thought it is done that way because of database performance issues when calling every source for every address, so to say `M x N` data accesses instead of two for loops.

So I'll change it to the function based / the one address db query approach. Unfortunately, I'm busy with other stuff the next days, so it'll take some time to change that. Just in case you're wondering :)

LefterisJP replied:
> 1. Style: For the style issue, is there a definition on when to split imports on multiple lines? I now run the `isort` with the parameters `--wrap-length=99 --line-length=99 --multi-line=3` to have it the way it is styled in other places but it won't catch the 99 columns long import in the `test_ethereum_names.py`.

For imports isort is fine. It's one of the few things we have completely automated.


> 3. Iterating over the adresses and query every source for that specific address was indeed my first approach. But I took a look at the previous implementation where each data source is queried with all addresses and I thought it is done that way because of database performance issues when calling every source for every address, so to say `M x N` data accesses instead of two for loops.

No there is no such concern.

tomvodi commented:
For this "initially wrong label" issue, is this some stuff to tackle in another issue, or should I fix it here? My preference would be to fix it right here. 
My first idea was to return from the [getting-blockchain-account-data](https://rotki.readthedocs.io/en/latest/api.html#getting-blockchain-account-data) endpoint right away with the prioritized name. In this case, the frontend shouldn't need any change. On the second glance, it makes the API harder to understand for the editing-blockchain-account-data endpoint for example, where the API user must know about which name is changed (most probably not the one the user has prioritized).

So, I would implement the 2. approach mentioned by @kelsos after I'm done with the other stuff.

tomvodi commented:
@LefterisJP @kelsos  I've fixed all open issues.
With my last commit I've also fixed the initial display of the names, though I'm not fully satisfied with that solution. If I click on Accounts & Balances, the user defined name shows up but only after some time waiting for it. Maybe there is a better solution to that. @kelsos do you have a better approach to that?

kelsos commented on frontend/app/src/store/session/index.ts:
```diff
@@ -102,6 +104,7 @@ export const useSessionStore = defineStore('session', () => {
       fetchNetValue()
     ]);
     await refreshPrices();
+    await refreshBlockchainBalances();
```
If you follow `fetch` you can see that it already does get accounts/blockchain balances. By adding refresh you will force one more refresh to happen really close to the previous one.

tomvodi replied:
I'll remove that line

kelsos commented on frontend/app/src/components/helper/PrioritizedList.vue:
```diff
@@ -155,6 +185,13 @@ const isLast = (item: string): boolean => {
+  identifier: PrioritizedListId
+): PrioritizedListItemData<PrioritizedListId> => {
+  const data = get(allItems);
+  return data.itemDataForId(identifier) ?? { identifier: PriceOracle.UNISWAP2 }; // TODO: make some NULL type
```
Please don't forget about the TODO here 😄

kelsos commented on frontend/app/src/components/helper/PrioritizedListEntry.vue:
```diff
@@ -0,0 +1,81 @@
+  [AddressNamePriority.PRIVATE_ADDRESSBOOK]: t(
+    'eth_address_book.hint.priority.list.private_address_book'
+  ).toString(),
+  blockchain: '',
+  coingecko: '',
+  cryptocompare: '',
+  fiat: '',
+  manual: '',
+  saddle: '',
+  defillama: ''
```
We don't have labels for these ones hmm.
For some, it doesn't really make sense to have translatable resources.

Anyway, it looks way better than before thank you 😄

tomvodi replied:
These keys don't have labels as they are only used as they are but converted to sentence case. Because of the code `{ [keys in PrioritizedListId]: string }` I get an "property ... is missing in type" error when I leave them out.

kelsos commented:
> @LefterisJP @kelsos I've fixed all open issues.
> With my last commit I've also fixed the initial display of the names, though I'm not fully satisfied with that solution. If I click on Accounts & Balances, the user defined name shows up but only after some time waiting for it. Maybe there is a better solution to that. @kelsos do you have a better approach to that?

We probably will have to add it on the initial loading, to avoid the situation.
Sorry for taking a bit to re-review, I was on vacation last week.

I left you a few minor comments mostly but overall it looks good. I only have to do a small functionality re-check and we should be gtg.

@LefterisJP @yabirgb is the backend part ok? can you have a look?

kelsos commented:
I did check the functionality quickly and it looks fine, the initial load if it is not there we can also fix it, since it was a bug that existed before the PR.

tomvodi commented:
That would be great if you could fix that initial loading.

LefterisJP commented on rotkehlchen/api/v1/schemas.py:
```diff
@@ -955,6 +955,7 @@ class ModifiableSettingsSchema(Schema):
     )
     cost_basis_method = SerializableEnumField(enum_class=CostBasisMethod, load_default=None)
     eth_staking_taxable_after_withdrawal_enabled = fields.Boolean(load_default=None)
+    address_name_priority = fields.List(fields.String(), load_default=None)
```
That's not what I meant. You can actually do validation as part ofthe fields.String() constructor.

An example from our code:

```python
    sync_approval = fields.String(
        load_default='unknown',
        validate=webargs.validate.OneOf(choices=('unknown', 'yes', 'no')),
    )
    ```

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -46,7 +45,8 @@ def find_ens_mappings(
                     addresses_to_query.append(address)
                 elif has_name:
-                    ens_mappings[cached_value.address] = cached_value.name  # type: ignore  # mypy doesn't see `isinstance` check  # noqa: E501
+                    ens_mappings[
+                        cached_value.address] = cached_value.name  # type: ignore  # mypy doesn't see `isinstance` check  # noqa: E501
```
we never break lines like that. And since you already got the noqa: E501 this is doubly not needed

```suggestion
                    ens_mappings[cached_value.address] = cached_value.name  # type: ignore  # mypy doesn't see `isinstance` check  # noqa: E501
```

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+            for name_source in prioritized_name_source:
+                fetcher = self._fetchers.get(name_source)
+                if not fetcher:
+                    raise NotImplementedError(
+                        f"address name fetcher for '{name_source}' is not implemented")
```
Don't split it like this. Closing parentheses at new line, or entire thing in one line. Finally we use `'` quotes.

```suggestion
                    raise NotImplementedError(
                        f'address name fetcher for "{name_source}" is not implemented',
                    )
```

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+
+
+def blockchain_address_to_name(db: DBHandler, address: ChecksumEvmAddress) -> str:
+    with db.conn.read_ctx() as cursor:
```
1. Please add a docstring

2. That's not a good approach. What you are doing here is you query all the account data and its labels from the DB. Then you iterate them and if you find the `address` argument then you return it. Otherwise you return empty string.
What you should do instead is do a direct query in the DB for this address and get the label if existing. IF not return `None`.

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+def private_addressbook_address_to_name(db: DBHandler, address: ChecksumEvmAddress) -> str:
+    db_addressbook = DBAddressbook(db)
+    with db.conn.read_ctx() as cursor:
+        addressbook_entries = db_addressbook.get_addressbook_entries(
```
same comment as above. Perhaps make this query as a method of `DBAddressbook`

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+
+def global_addressbook_address_to_name(db: DBHandler, address: ChecksumEvmAddress) -> str:
+    db_addressbook = DBAddressbook(db)
+    with db_addressbook.read_ctx(AddressbookType.GLOBAL) as global_cursor:
```
same here

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+
+
+def token_mappings_address_to_name(_: DBHandler, address: ChecksumEvmAddress) -> str:
+    token_mappings = GlobalDBHandler().get_tokens_mappings(addresses=[address])
+
+    for _addr, name in token_mappings.items():
+        return name
+
+    return ""
```
1. Add a docstring.
2. You are asking for a single address and get the mapping for this address. Why are you iterating the mapping and return the first thing you get?

```suggestion
    token_mappings = GlobalDBHandler().get_tokens_mappings(addresses=[address])
    return token_mappings.get(address, None)
```

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
+    return ""
+
+
+def ens_address_to_name(db: DBHandler, address: ChecksumEvmAddress) -> str:
```
add a docstring

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -67,56 +67,135 @@ def search_for_addresses_names(
-        result_mappings[addressbook_entry.address] = addressbook_entry.name
-    for addr, label in labels.items():
-        if label is not None:
-            result_mappings[to_checksum_address(addr)] = label
-    for addressbook_entry in private_addressbook:
-        result_mappings[addressbook_entry.address] = addressbook_entry.name
-
-    return result_mappings
+        if isinstance(mapping, EnsMapping) and mapping.name:
+            return mapping.name
+
+    return ""
```
What are you doing here? You are taking a mapping of 1 value. The value you need. Then you turn that into a list? Why? And then iterate that list? And check that the list's type is what the list has and return the first item?!?!

Just do this:

```python
db_ens = DBEns(db)
with db.conn.read_ctx() as cursor:
    db_reverse_ens = db_ens.get_reverse_ens(cursor, addresses=[address])
    return db_reverse_ens.get(address, None)
```

tomvodi replied:
As far as I understand, at least the `isinstance` check must be there before returning a value as the `get_reverse_ens` method may return a `Timestamp`

LefterisJP replied:
Hmmm that's true. Then you can keep the is_instance there.

But then the `get_reverse_ens` is terribly designed by whoever made it.

We would need to change it to only return mapped names, and have a different named function that also returns timestamp. But that can be done in another PR.

LefterisJP commented on rotkehlchen/tests/unit/test_ethereum_names.py:
```diff
@@ -0,0 +1,73 @@
+    fetchers: Dict[AddressNameSource, FetcherFunc] = {}
+    for source_id in name_source_ids:
+        def make_fetcher(src) -> FetcherFunc:
+            return lambda db, addr: f"{src} address name"
```
code style:

```suggestion
            return lambda db, addr: f'{src} address name'
```

LefterisJP commented on rotkehlchen/tests/unit/test_ethereum_names.py:
```diff
@@ -0,0 +1,73 @@
+        fetchers,
+) -> Dict[AddressNameSource, FetcherFunc]:
+    def empty_source(_db, _addr):
+        return ""
```
```suggestion
        return ''
```

LefterisJP commented on rotkehlchen/tests/unit/test_ethereum_names.py:
```diff
@@ -0,0 +1,73 @@
+    return fetchers
+
+
+def test_get_prioritized_name(name_source_ids, fetchers, evm_address):
```
Write a docstring here and in the next test explaining what you are testing. It's a small test but very hard to follow

tomvodi replied:
I'll rewrite the tests to be easier to follow.

LefterisJP reviewed (commented):
Lots of comments.

Also some tests failed due to the PR changes here and need adjustment:

- rotkehlchen/tests/api/test_settings.py::test_set_settings
- rotkehlchen/tests/db/test_db.py::test_writing_fetching_data
- rotkehlchen/tests/unit/test_pnl_debug.py::test_pnl_debug_import 

I am not sure I understand the tests you wrote to be perfectly honest also :sweat_smile:

LefterisJP commented on rotkehlchen/api/v1/schemas.py:
```diff
@@ -975,6 +976,9 @@ class ModifiableSettingsSchema(Schema):
     eth_staking_taxable_after_withdrawal_enabled = fields.Boolean(load_default=None)
+    address_name_priority = fields.List(fields.String(
+        validate=webargs.validate.OneOf(choices=DEFAULT_ADDRESS_NAME_PRIORITY),
+    ), load_default=None)
```
Why have `load_default=None` here? You can actually use the default priority as the default choice if nothing is given. This way the argument won't be optional at all.

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -65,58 +64,146 @@ def find_ens_mappings(
+
+
+class NamePrioritizer:
+    _fetchers: Dict[AddressNameSource, FetcherFunc] = {}
+    _db: DBHandler
+
+    def __init__(self, database: DBHandler):
+        self._db = database
```
Don't put attributes at the class level. These are going to be for every instance of name prioritize which is definitely not what you want. They should be in the `init()`

```suggestion

    def __init__(self, database: DBHandler):
        self._fetchers: Dict[AddressNameSource, FetcherFunc] = {}
        self._db = database
```

LefterisJP commented on rotkehlchen/chain/ethereum/names.py:
```diff
@@ -65,58 +64,146 @@ def find_ens_mappings(
+    None if there is no such account or the account has no label set.
+    """
+    with db.conn.read_ctx() as cursor:
+        return db.get_blockchain_account_label(
```
nitpick for this and the rest. We like kwargs in rotki codebase, since this:

1. won't error due to mistaken argument order
2. python interpreter will actually warn you if the argument does not exist even at parsing time

LefterisJP commented on rotkehlchen/db/addressbook.py:
```diff
@@ -110,3 +110,20 @@ def delete_addressbook_entries(
+            book_type: AddressbookType,
+            address: ChecksumEvmAddress,
+    ) -> Optional[str]:
+        with self.write_ctx(book_type) as read_cursor:
```
If it's a read cursor then you should use a different context.
```suggestion
        with self.read_ctx(book_type) as read_cursor:
```

LefterisJP commented on rotkehlchen/db/addressbook.py:
```diff
@@ -110,3 +110,20 @@ def delete_addressbook_entries(
+            self,
+            book_type: AddressbookType,
+            address: ChecksumEvmAddress,
+    ) -> Optional[str]:
```
add docstring

LefterisJP commented on rotkehlchen/db/dbhandler.py:
```diff
@@ -3449,3 +3450,25 @@ def get_nft_mappings(self, identifiers: List[str]) -> Dict[str, dict]:
+            (blockchain.value, address),
+        )
+        result = query.fetchone()
+        if result:
+            return result[0]
+
+        return None
```
can shorten

```suggestion
        return None if result is None else result[0]
```

LefterisJP commented on rotkehlchen/db/addressbook.py:
```diff
@@ -110,3 +110,20 @@ def delete_addressbook_entries(
+            )
+
+            result = query.fetchone()
+            if result:
+                return result[0]
+
+        return None
```
can shorten

```suggestion
        return None if result is None else result[0]
```

LefterisJP commented on rotkehlchen/db/dbhandler.py:
```diff
@@ -3449,3 +3450,25 @@ def get_nft_mappings(self, identifiers: List[str]) -> Dict[str, dict]:
+        """Returns the label for a specific blockchain account.
+        Returns None if either there is no label set or the account doesn't exist in database.
+        """
+        query = cursor.execute(
+            'SELECT label '
+            'FROM blockchain_accounts '
+            'WHERE blockchain=? '
+            'AND account=?',
+            (blockchain.value, address),
+        )
```
can reduce lines number
```suggestion
        query = cursor.execute(
            'SELECT label FROM blockchain_accounts '
            'WHERE blockchain=? AND account=?',
            (blockchain.value, address),
        )
```

LefterisJP reviewed (approved):
Hey @tomvodi looks much better. Left few smaller comments, mostly nitpicks except one. Still none is breaking. I would like to merge this PR to allow the work to continue on this.

You can either fix what I asked in another PR or leave it to us. Let me know.
